File: app/src/main/java/com/example/puyo_base_simulator/ui/home/HomePresenter.java
package com.example.puyo_base_simulator.ui.home;

import android.app.Activity;
import android.content.res.AssetManager;

import androidx.room.Room;

import com.example.puyo_base_simulator.data.AppDatabase;
import com.example.puyo_base_simulator.data.Base;

import org.apache.commons.lang.SerializationUtils;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Serializable;
import java.util.Random;

interface ButtonUpdateFunction {
    void func();
}

public class HomePresenter implements HomeContract.Presenter {
    Activity mActivity;
    Field currentField;
    StackWithButton<Field> fieldStack;
    StackWithButton<Placement> fieldRedoStack;
    TsumoController tsumoController;
    private static final Random RANDOM = new Random();
    HomeFragment mView;
    AppDatabase mDB;
    Haipuyo mHaipuyo = Haipuyo.getInstance();

    HomePresenter(HomeFragment view, AssetManager asset, Activity activity) {
        mDB = Room.databaseBuilder(activity.getApplicationContext(),
              AppDatabase.class, "database-name")
              .allowMainThreadQueries() // Main thread でも動作させたい場合
              .build();
        mView = view;
        mActivity = activity;
        try {
            InputStream haipuyoIs = asset.open("haipuyo.txt");
            BufferedReader haipuyoBr = new BufferedReader(new InputStreamReader(haipuyoIs));
            InputStream sortedIs = asset.open("sorted_haipuyo.txt");
            BufferedReader sortedBr = new BufferedReader(new InputStreamReader(sortedIs));
            mHaipuyo.load(haipuyoBr, sortedBr);
        } catch (IOException e) {
            e.printStackTrace();
        }
        currentField =  new Field(1);
        fieldStack = new StackWithButton<>(() -> mView.enableUndoButton(), () -> mView.disableUndoButton());
        fieldRedoStack = new StackWithButton<>(() -> mView.enableRedoButton(), () -> mView.disableRedoButton());
        int seed = RANDOM.nextInt(65536);
        tsumoController = new TsumoController(mHaipuyo.get(seed), seed);
        mView.setSeedText(tsumoController.seed);
        mView.update(currentField, tsumoController.makeTsumoInfo());
    }

    public void rotateLeft() {
        tsumoController.rotateCurrentRight();
        TsumoInfo tsumoInfo = tsumoController.makeTsumoInfo();
        mView.update(currentField, tsumoInfo);
    }

    public void rotateRight() {
        tsumoController.rotateCurrentLeft();
        TsumoInfo tsumoInfo = tsumoController.makeTsumoInfo();
        mView.update(currentField, tsumoInfo);
    }

    public void moveLeft() {
        tsumoController.moveCurrentLeft();
        TsumoInfo tsumoInfo = tsumoController.makeTsumoInfo();
        mView.update(currentField, tsumoInfo);
    }

    public void moveRight() {
        tsumoController.moveCurrentRight();
        TsumoInfo tsumoInfo = tsumoController.makeTsumoInfo();
        mView.update(currentField, tsumoInfo);
    }

    private Field setPairOnField() {
        Field newField = (Field) SerializationUtils.clone(currentField);
        Rotation currentCursorRotate = tsumoController.currentCursorRotate;
        int currentCursorColumnIndex = tsumoController.currentCursorColumnIndex;
        boolean success = true;
        switch (currentCursorRotate) {
            case DEGREE0:
                // jiku puyo
                success = newField.addPuyo(currentCursorColumnIndex, tsumoController.getMainColor());
                // non-jiku puyo
                success &= newField.addPuyo(currentCursorColumnIndex, tsumoController.getSubColor());
                break;
            case DEGREE90:
                // jiku puyo
                success = newField.addPuyo(currentCursorColumnIndex, tsumoController.getMainColor());
                // non-jiku puyo
                success &= newField.addPuyo(currentCursorColumnIndex + 1, tsumoController.getSubColor());
                break;
            case DEGREE180:
                // 上下が逆転している
                // non-jiku puyo
                success = newField.addPuyo(currentCursorColumnIndex, tsumoController.getSubColor());
                // jiku puyo
                success &= newField.addPuyo(currentCursorColumnIndex, tsumoController.getMainColor());
                break;
            case DEGREE270:
                // jiku puyo
                success = newField.addPuyo(currentCursorColumnIndex, tsumoController.getMainColor());
                // non-jiku puyo
                success &= newField.addPuyo(currentCursorColumnIndex - 1, tsumoController.getSubColor());
                break;
        }
        if (success) {
            return newField;
        } else {
            return null;
        }
    }

    public void dropDown() {
        Field newFiled = setPairOnField();
        if (newFiled == null) {
            return;
        }
        fieldStack.push(currentField);
        currentField = newFiled;
        tsumoController.pushPlacementOrder();
        fieldRedoStack.clear();
        mView.drawField(currentField);
        currentField.evalNextField();
        tsumoController.incrementTsumo();
        if (currentField.nextField == null) {
            mView.drawTsumo(tsumoController.makeTsumoInfo(), currentField);
        } else {
            mView.eraseCurrentPuyo();
            mView.disableAllButtons();
            drawFieldChain(currentField);
            currentField = getLastField(currentField);
        }
    }

    public void undo() {
        fieldRedoStack.push(tsumoController.popPlacementOrder());
        currentField = fieldStack.pop();
        tsumoController.decrementTsumo();
        TsumoInfo tsumoInfo = tsumoController.makeTsumoInfo();
        mView.update(currentField, tsumoInfo);
    }

    public void redo() {
        tsumoController.restorePlacement(fieldRedoStack.pop());
        fieldStack.push(currentField);
        currentField = setPairOnField();
        tsumoController.pushPlacementOrder();
        mView.drawField(currentField);
        currentField.evalNextField();
        tsumoController.incrementTsumo();
        if (currentField.nextField == null) {
            mView.drawTsumo(tsumoController.makeTsumoInfo(), currentField);
        } else {
            mView.eraseCurrentPuyo();
            mView.disableAllButtons();
            drawFieldChain(currentField);
            currentField = getLastField(currentField);
        }
    }

    public void save() {
        Base base = new Base();
        base.setHash(tsumoController.seed);
        base.setField(currentField.toString());
        base.setPlacementOrder(tsumoController.placementOrderToString());
        mDB.baseDao().insert(base);
    }

    public void load(FieldPreview fieldPreview) {
        Base base = mDB.baseDao().findById(fieldPreview.id);
        currentField = new Field(1);
        tsumoController.stringToPlacementOrder(base.getPlacementOrder());
        fieldRedoStack.clear();
        while (!tsumoController.placementOrder.isEmpty()) {
            fieldRedoStack.push(tsumoController.popPlacementOrder());
        }
        fieldStack.clear();
        tsumoController = new TsumoController(mHaipuyo.get(base.getHash()), base.getHash());
        mView.update(currentField, tsumoController.makeTsumoInfo());

    }

    Field getLastField(Field field) {
        if (field.nextField == null) {
            return field;
        } else {
            return getLastField(field.nextField);
        }
    }


    void drawFieldChain(final Field field) {
        drawFieldChainRecursive(field, true);
    }

    void drawFieldChainRecursive(final Field field, boolean disappear) {
        new Thread(() -> {
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            if (disappear) {
                String text = "" + field.bonus + " * " + field.disappearPuyo.size() + " = " + field.accumulatedPoint + "点";
                mView.drawPoint(text);
                mActivity.runOnUiThread(() -> mView.drawDisappearField(field));
                drawFieldChainRecursive(field.nextField, false);
            } else {
                mActivity.runOnUiThread(() -> mView.drawField(field));
                if (field.disappearPuyo.isEmpty()) {
                    // 終了処理
                    mActivity.runOnUiThread(() -> {
                        mView.enableAllButtons();
                        if (fieldRedoStack.isEmpty()) {
                            mView.disableRedoButton();
                        }
                        TsumoInfo tsumoInfo = tsumoController.makeTsumoInfo();
                        mView.update(field, tsumoInfo);
                    });
                } else {
                    drawFieldChainRecursive(field, true);
                }
            }
        }).start();
    }

    public void setSeed() {
        try {
            int newSeed = mView.getSpecifiedSeed();
            tsumoController = new TsumoController(mHaipuyo.get(newSeed), newSeed);
            fieldRedoStack.clear();
            fieldStack.clear();
            mView.setSeedText(newSeed);
            currentField =  new Field(1);
            mView.update(currentField, tsumoController.makeTsumoInfo());
        } catch (NumberFormatException ignored) {

        }
    }
}

File: app/src/main/java/com/example/puyo_base_simulator/ui/home/HomePresenter.kt
package com.example.puyo_base_simulator.ui.home

import android.app.Activity
import android.content.Context
import android.content.res.AssetManager
import androidx.compose.ui.text.input.TextFieldValue
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.room.Room
import com.example.puyo_base_simulator.data.AppDatabase
import com.example.puyo_base_simulator.data.Base
import org.apache.commons.lang.SerializationUtils
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.util.*
import java.util.stream.Collectors
import kotlin.math.roundToInt


class HomePresenter internal constructor(asset: AssetManager) : ViewModel() {
    var fieldHistory = History<Field>()
    private var tsumoController: TsumoController
    private var mDB: AppDatabase? = null
    // todo: lateinitとか何かで消せない？
    private val sampleTsumoInfo = TsumoInfo(
        Array(2) {PuyoColor.RED},
        Array(2) {Array(2) {PuyoColor.RED}},
        3,
        Rotation.DEGREE0
    )
    private val _tsumoInfo = MutableLiveData(sampleTsumoInfo)
    val tsumoInfo: LiveData<TsumoInfo> = _tsumoInfo
    private val _currentField = MutableLiveData(Field())
    val currentField: LiveData<Field> = _currentField
    private val _seed = MutableLiveData(0)
    val seed: LiveData<Int> = _seed
    private val _historySliderValue = MutableLiveData(0f)
    val historySliderValue = _historySliderValue
    private val _historySize = MutableLiveData(0)
    val historySize = _historySize

    private fun getDB(context: Context) : AppDatabase {
        if (mDB == null) {
            val result = Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java, "database-name"
            ).allowMainThreadQueries()
                    .build()
            mDB = result
        }
        return mDB!!
    }

    fun rotateLeft() {
        tsumoController.rotateCurrentRight()
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
    }

    fun rotateRight() {
        tsumoController.rotateCurrentLeft()
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
    }

    fun moveLeft() {
        tsumoController.moveCurrentLeft()
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
    }

    fun moveRight() {
        tsumoController.moveCurrentRight()
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
    }

    private fun setPairOnField(field: Field, tsumoInfo: TsumoInfo): Field? {
        val col = tsumoInfo.column
        val mainColor = tsumoInfo.currentColor[0]
        val subColor = tsumoInfo.currentColor[1]
        val newField = SerializationUtils.clone(field) as Field
        val success = when (tsumoInfo.rot) {
            Rotation.DEGREE0 -> newField.addPuyo(col, mainColor) and newField.addPuyo(col, subColor)
            Rotation.DEGREE90 -> newField.addPuyo(col, mainColor) and newField.addPuyo(col + 1, subColor)
            Rotation.DEGREE180 -> newField.addPuyo(col, subColor) and newField.addPuyo(col, mainColor) // 上下が逆転している
            Rotation.DEGREE270 -> newField.addPuyo(col, mainColor) and newField.addPuyo(col - 1, subColor)
        }
        return if (success) newField else null
    }

    fun dropDown(activity: Activity) : Field? {
        val newField = setPairOnField(currentField.value!!, tsumoController.makeTsumoInfo()) ?: return null
        tsumoController.addPlacementHistory()
        newField.evalNextField()
        _currentField.value = newField
        fieldHistory.add(newField)
        _historySliderValue.value = fieldHistory.index.toFloat()
        _historySize.value = fieldHistory.size()
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
        if (newField.nextField != null) {
            chain(newField, activity)
        }
        return newField
    }

    fun undo() {
        tsumoController.undoPlacementHistory()
        _currentField.value = fieldHistory.undo()!!
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
    }

    fun redo(activity: Activity) : Field {
        fieldHistory.redo()
        _historySliderValue.value = fieldHistory.index.toFloat()
        val field = setPairOnField(currentField.value!!, tsumoController.makeTsumoInfo(tsumoController.currentPlacementHistory()))!!
        tsumoController.redoPlacementHistory()
        field.evalNextField()
        _currentField.value = field
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
        if (field.nextField != null) {
            chain(field, activity)
        }
        return field
    }

    fun save(context: Context) : Boolean {
        if (fieldHistory.isFirst()) return false
        val base = Base()
        base.hash = seed.value!!
        base.placementHistory = tsumoController.placementOrderToString()
        val field = currentField.value!!
        base.allClear = field.allClear()
        base.point = field.accumulatedPoint
        base.field = if (field.allClear()) {
            fieldHistory.previous().toString()
        } else {
            field.toString()
        }
        getDB(context).baseDao().insert(base)
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
        return true
    }

    fun load(base: Base) {
        tsumoController = TsumoController(Haipuyo[base.hash], base.hash)
        clearFieldHistory()
        var f = Field()
        for (p in tsumoController.stringToPlacementOrder(base.placementHistory)) {
            f = setPairOnField(f, tsumoController.makeTsumoInfo(p))!!
            f.evalNextField()
            f = getLastField(f)
            fieldHistory.add(f)
        }
        tsumoController.addPlacementHistory()
        tsumoController.rollbackPlacementHistory()
        _currentField.value = fieldHistory.undoAll()
        _historySliderValue.value = fieldHistory.index.toFloat()
        _historySize.value = fieldHistory.size()
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
        _seed.value = base.hash // TODO: 名前がおかしいので修正する
    }

    private fun clearFieldHistory() {
        fieldHistory.clear()
        fieldHistory.add(Field())
        _historySliderValue.value = fieldHistory.index.toFloat()
        _historySize.value = fieldHistory.size()
    }

    private fun getLastField(field: Field): Field {
        val next = field.nextField
        return if (next == null) {
            field
        } else {
            getLastField(next)
        }
    }

    fun setSeed(newSeed: Int) {
        _seed.value = newSeed
        tsumoController = TsumoController(Haipuyo[newSeed], newSeed)
        _currentField.value = Field()
        clearFieldHistory()
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
    }

    fun randomGenerate() {
        _currentField.value = Field()
        clearFieldHistory()
        val seed = RANDOM.nextInt(65536)
        _seed.value = seed
        tsumoController = TsumoController(Haipuyo[seed], seed)
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
    }

    fun setHistoryIndex(idx: Float) {
        _historySliderValue.value = idx
        fieldHistory.index = idx.roundToInt()
        _currentField.value = fieldHistory.current()
        tsumoController.setHistoryIndex(idx.roundToInt())
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
    }

    private fun chain(field: Field, activity: Activity) {
        Thread {
            Thread.sleep(500)
            activity.runOnUiThread {
                _currentField.value = field.disappearingField
            }
            Thread.sleep(500)
            activity.runOnUiThread {
                _currentField.value = field.nextField!!
            }
            if (field.nextField!!.nextField != null) {
                chain(field.nextField!!, activity)
            }
        }.start()
    }

    fun searchBySeed(seed : Int, context: Context) : MutableList<Base> {
        if (seed !in 0..65535) {
            throw NumberFormatException("should enter 0-65535")
        }
        val bases = getDB(context).baseDao().findByHash(seed)
        return bases.toMutableList()
    }

    fun searchByPattern(pattern : String, context: Context) : MutableList<Base> {
        val seeds = Haipuyo.searchSeedWithPattern(pattern)
        val seedsChunks = seeds.chunked(100)
        val bases = seedsChunks.parallelStream().map { seed: List<Int> -> getDB(context).baseDao().findByAllHash(seed) }.flatMap { obj: List<Base> -> obj.stream() }.collect(
            Collectors.toList())
        return bases.toMutableList()
    }

    fun showAll (context: Context) : MutableList<Base> {
        val bases = getDB(context).baseDao().all
        return bases.toMutableList()
    }

    companion object {
        private val RANDOM = Random()
    }

    init {
        try {
            val haipuyoIs = asset.open("haipuyo.txt")
            val haipuyoBr = BufferedReader(InputStreamReader(haipuyoIs))
            val sortedIs = asset.open("sorted_haipuyo.txt")
            val sortedBr = BufferedReader(InputStreamReader(sortedIs))
            Haipuyo.load(haipuyoBr, sortedBr)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        val seed = RANDOM.nextInt(65536)
        _seed.value = seed
        tsumoController = TsumoController(Haipuyo[seed], seed)
        _tsumoInfo.value = tsumoController.makeTsumoInfo()
        clearFieldHistory()
    }
}